File: node-martian-ddd/src/domain/navigator.js
"use strict";
const { contains } = require("./grid");

function execute(scents, grid, robot, instr) {
  for (const c of instr) {
    if (robot.lost) break;
    if (c === "L") robot.turnLeft();
    else if (c === "R") robot.turnRight();
    else if (c === "F") {
      const next = robot.nextForward();
      if (!contains(grid.maxX, grid.maxY, next)) {
        const keyPos = robot.position;
        const keyO = robot.orientation;
        if (!scents.has(keyPos, keyO)) {
          scents.add(keyPos, keyO);
          robot.lost = true;
        }
      } else {
        robot.position = next;
      }
    }
  }
}

module.exports = { execute };

File: langs/csharp-ddd/src/MartianDDD.Domain/Navigator.cs
namespace MartianDDD.Domain;

public sealed class Navigator
{
    private readonly IScentRepository _scents;

    public Navigator(IScentRepository scents)
    {
        _scents = scents;
    }

    public void Execute(Grid grid, Robot robot, string instructions)
    {
        foreach (var ch in instructions)
        {
            if (robot.Lost) break;
            switch (ch)
            {
                case 'L': robot.TurnLeft(); break;
                case 'R': robot.TurnRight(); break;
                case 'F':
                    var next = robot.NextForward();
                    if (!grid.Contains(next))
                    {
                        // Check scent at current position+orientation. If present, ignore the move.
                        if (!_scents.HasScent(robot.Position, robot.Orientation))
                        {
                            _scents.AddScent(robot.Position, robot.Orientation);
                            robot.MarkLost();
                        }
                    }
                    else
                    {
                        robot.MoveTo(next);
                    }
                    break;
            }
        }
    }
}
